### PR TITLE
Updating xUnit to v2.1.0 Beta3

### PR DESCRIPTION
**Pull request by tannergooding**

### CODE REVIEW

**Comment by KevinH-MS** [comment]
No test results at all--oh my! :fearful:

**Comment by tannergooding** [comment]
FYI. @KevinH-MS, @agocke 

New commit that adds the WpfFactAttribute, WpfFactDiscoverer, WpfTestCase, WpfTheoryAttribute, and WpfTheoryDiscoverer types. These are based on the samples provided by xUnit (see xunit/samples.xunit -> STAExamples).

**Comment by tannergooding** [comment]
@dotnet-bot test this please

**Comment by agocke** [comment]
@tannergooding Could you change your individual commits to have their summary just be the change that was made and not include "Updating xUnit to v2.1.0 Beta3"? The PR merge commit should contain that information and it's harder to just eyeball what each commit does if the change isn't on the first line.

**Comment by tannergooding** [comment]
FYI. @jasonmalinowski, @jaredpar This may end up impacting `Microsoft.CodeAnalysis.Editor.Shared.Utilities.ForegroundThreadAffinitizedObject`.

**Comment by agocke** [comment]
Btw i checked yesterday and beta4 works with td.net

**Comment by KevinH-MS** [comment]
It we can, it'd be great to move to beta4 then (so td.net works out of the box).

**Comment by tannergooding** [comment]
Talked with @jaredpar offline, and because of how our tests are configured and were originally written, we can upgrade provided we pass `-parallel none` to xunit. We still need to have the new `WpfFact` to explicitly run the tests on an STA Thread and long term we will need to fix `ForegroundThreadAffinitizedObject` so that we can enable the in process parallelization that xunit performs.

**Comment by tannergooding** [comment]
@KevinH-MS, @agocke. I am working on updating the PR to beta4 and fixing the few remaining issues blocking (workaround for `ForegroundThreadAffinitizedObject` being broken).

**Comment by KevinH-MS** [comment]
For now, rather than -parallel none, can we just turn off the parallelism with an assembly attribute (for the test assemblies that are affected)?  This will prevent issues with other runners.

**Comment by jaredpar** [comment]
@KevinH-MS that's a great idea.  It also gives us the flexibility to opt specific assemblies into parallelization.

**Comment by tannergooding** [comment]
@KevinH-MS, @jaredpar. I will fix it to use attributes rather than passing `-parallel none`.

**Comment by tannergooding** [comment]
@KevinH-MS, @Pilchie, @jaredpar, @agocke, @ManishJayaswal 

I believe all the issues have been resolved now. Only the tests which actually use WPF run using WpfFact, all tests are run in serial (until we can fix the ForegroundThreadAffinitizedObject issues), all tests are configured via the App.config file (which allows testing/configuration without rebuild).

**Comment by agocke** [comment]
As mentioned offline, we need copyright notices for the xunit WPF sample stuff, but otherwise :+1:

**Comment by tannergooding** [comment]
Added the copyright notices for WpfFact (and friends) from OuterCurve Foundation (http://github.com/xunit/samples.xunit).

File: BuildAndTest.proj [hunk 1]
```diff
@@ -3,10 +3,6 @@
   DefaultTargets="BuildAndTest"
   ToolsVersion="12.0">
 
-  <UsingTask
-    AssemblyFile="packages\xunit.runners.2.0.0-alpha-build2576\tools\xunit.runner.msbuild.dll"
-    TaskName="Xunit.Runner.MSBuild.xunit" />
-
   <!-- At the moment Jenkins runs on a machine without any build of Visual Studio 2015.  Hence we must
```
Comment (jasonmalinowski) [inline BuildAndTest.proj]: Stale comment.

File: BuildAndTest.proj [hunk 2]
```diff
@@ -48,7 +43,7 @@
       <TestAssemblies Include="Binaries\$(Configuration)\**\$(IncludePattern)" />
     </ItemGroup>
 
-    <Exec Command="Binaries\$(Configuration)\RunTests.exe packages\xunit.runners.2.0.0-alpha-build2576\tools $(RunTestArgs) @(TestAssemblies, ' ')" />
+    <Exec Command="Binaries\$(Configuration)\RunTests.exe packages\xunit.runner.console.2.1.0-beta4-build3109\tools $(RunTestArgs) @(TestAssemblies, ' ')" />
```
Comment (jasonmalinowski) [inline BuildAndTest.proj]: Any chance we can avoid the version hardcode? We've been bitten a lot by NuGet versions being baked in many places and not updating all of them.

File: src/Compilers/CSharp/Test/CommandLine/app.config [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<configuration>
+  <appSettings>
+    <add key="xunit.useAppDomain" value="true"/>
+    <add key="xunit.diagnosticMessages" value="false"/>
```
Comment (jasonmalinowski) [inline src/Compilers/CSharp/Test/CommandLine/app.config]: Per http://xunit.github.io/docs/configuring-with-xml.html, these first two are the default? Can we delete them?

File: src/Compilers/Core/VBCSCompiler/ServerDispatcher.cs [hunk 4]
```diff
@@ -147,8 +147,6 @@ public ServerDispatcher(IRequestHandler handler, IDiagnosticListener diagnosticL
             Justification = "We intentionally call GC.Collect when anticipate long period on inactivity.")]
         public void ListenAndDispatchConnections(string pipeName, TimeSpan? keepAlive, CancellationToken cancellationToken = default(CancellationToken))
         {
-            Debug.Assert(SynchronizationContext.Current == null);
```
Comment (jasonmalinowski) [inline src/Compilers/Core/VBCSCompiler/ServerDispatcher.cs]: This looks like a pretty important assert. Can we preserve it?

**Comment by jasonmalinowski** [comment]
:+1:. I see you've added the licensing information, I trust you're talking to the lawyers to make sure that's good.

**Comment by agocke** [comment]
@tannergooding Seconding what @jasonmalinowski says -- we should loop in legal to make sure we're doing this right. Also, could you move the third-party folder to be a directory in the root of the repository? I want to make it *very* clear that this is licensed external code.

File: src/Compilers/Core/VBCSCompiler/ServerDispatcher.cs (hunk 4 above)
Comment (tannergooding) [inline src/Compilers/Core/VBCSCompiler/ServerDispatcher.cs]: @jaredpar and @agocke  were the ones who said to remove this.

Comment (agocke) [inline src/Compilers/Core/VBCSCompiler/ServerDispatcher.cs]: @jasonmalinowski It's invalid. I don't think we can assert anything about the test environment's synchronization context.

**Comment by tannergooding** [comment]
Looks like I need to rebase again, but the latest gets rid of the sample StaFact from xUnit and implements our own Sta and WpfFacts (which use a custom TaskScheduler rather than creating a new thread per test). So it performs and scales better.

I've also done some of the work required to allow parallelization of the tests moving forward.

**Comment by tannergooding** [comment]
@KevinH-MS, @Pilchie, @jaredpar, @agocke, @ManishJayaswal, @jasonmalinowski, @brettfo, @amcasey 

Significantly updated, still going through the last few tests, but everything seems to be much smoother and not flaky now.

**Comment by amcasey** [comment]
:+1: Seems reasonable.  Someone else should review the synchronization context changes.

**Comment by agocke** [comment]
If this passes :+1:

**Comment by tannergooding** [comment]
test linux please

**Pull request closed without merging**